[core] memory monitor documentation (#29341)
ray-project/ray: doc/source/ray-core/scheduling/ray-oom-prevention.rst

> Ray Out-Of-Memory Prevention

Reviewer: ```suggestion
Out-Of-Memory Prevention
```
I think Ray is unnecessary

============================

If application tasks or actors consume a large amount of heap space, it can cause the node to run out of memory (OOM). When that happens, the operating system will start killing worker or raylet processes, disrupting the application. OOM may also stall metrics and if this happens on the head node, it may stall the :ref:`dashboard <ray-dashboard>` or other control processes and cause the cluster to become unusable.

In this section we will go over:

- What is the memory monitor and how it works

- How to enable and configure it

- How to use the memory monitor to detect and resolve memory issues

What is the memory monitor?
---------------------------

> The memory monitor is a component that runs on each node in your Ray cluster and is instantiated with a configured threshold.

Reviewer 2: Remove this paragraph? Looks like it got duplicated somehow.

The memory monitor periodically checks the memory usage of the different Ray processes on the node,
which includes the worker heap, the object store, and the raylet as described in :ref:`memory management <memory>`.

> The memory monitor is a component that runs within the raylet process on each node. It periodically checks the memory usage, which includes the worker heap, the object store, and the raylet as described in :ref:`memory management <memory>`. If the combined usage exceeds a configurable threshold the raylet will kill a task or actor process to free up memory and prevent Ray from failing.

Reviewer: Can you add a link to the whitepaper? (since some ppl woudln't know raylet)

.. note::

>     The memory monitor is in :ref:`alpha <api-stability-alpha>`. It is disabled by default and needs to be enabled by setting the environment variable ``RAY_memory_monitor_interval_ms`` to a value greater than zero when Ray starts. It is available on Linux and is tested with Ray running inside a container that is using cgroup v1. If you encounter issues when running the memory monitor outside of a container or the container is using cgroup v2, please :ref:`file an issue or post a question <oom-questions>`.

Reviewer: Maybe this should be on the top of the doc?

Author: Seems this is the top already for discussing memory monitor?

    For a node with ~33 GiB of RAM the raylet log prints:

    .. code-block:: bash

        $ grep memory_monitor.cc /tmp/ray/session_latest/logs/raylet.out

        (raylet) memory_monitor.cc: MemoryMonitor initialized with usage threshold at 32164429824 bytes (0.90 system memory), total system memory bytes: 35738255360

    On the other hand, if we set ``RAY_min_memory_free_bytes`` to a low value it will limit the amount of free memory reserved by the memory monitor. This helps limit the free memory for a node with large amounts of RAM.

    .. code-block:: bash

        RAY_memory_monitor_interval_ms=100 RAY_memory_usage_threshold_fraction=0.9 RAY_min_memory_free_bytes=1000000000 ray start --head

    For a node with ~33 GiB of RAM it should have a threshold close to 32GiB by leaving only 1000000000 bytes of free memory.

    .. code-block:: bash

        $ grep memory_monitor.cc /tmp/ray/session_latest/logs/raylet.out

        (raylet) memory_monitor.cc: MemoryMonitor initialized with usage threshold at 34738255360 bytes (0.97 system memory), total system memory bytes: 35738255360

> Using the Memory Monitor

Reviewer: Call it a Policy?

Author: I'll let @stephanie-wang  decide since she asked to create a usage section

Reviewer 2: I think this is fine as is.

------------------------

Retry policy
~~~~~~~~~~~~

> When a task or actor is killed by the memory monitor, it will retry using a separate retry counter based off of ``RAY_task_oom_retries`` instead of the typical retry that is :ref:`max_retries <task-fault-tolerance>` for Tasks and :ref:`max_restarts <actor-fault-tolerance>` for Actors. The number of memory monitor retries is the same between tasks and actors and the defaults is ``15``. The value should be passed as the environment variable ``RAY_task_oom_retries`` when starting Ray as well as the application.

Reviewer 2: ```suggestion
When a task or actor is killed by the memory monitor, it will retry using a separate retry counter based off of ``RAY_task_oom_retries`` instead of the typical number of retries specified by :ref:`max_retries <task-fault-tolerance>` for tasks and :ref:`max_restarts <actor-fault-tolerance>` for actors. The number of memory monitor retries is the same for tasks and actors and defaults to ``15``. To override this value, the environment variable ``RAY_task_oom_retries`` should be set when starting Ray as well as the application.
```

Verify the task was indeed executed twice via ``task_oom_retry``:

.. code-block:: bash

    $ grep -r "retries left" /tmp/ray/session_latest/logs/

    /tmp/ray/session_latest/logs/python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_60002.log:[2022-10-12 16:14:07,723 I 60002 60031] task_manager.cc:458: task c8ef45ccd0112571ffffffffffffffffffffffff01000000 retries left: 3, oom retries left: 1, task failed due to oom: 1

    /tmp/ray/session_latest/logs/python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_60002.log:[2022-10-12 16:14:18,843 I 60002 60031] task_manager.cc:458: task c8ef45ccd0112571ffffffffffffffffffffffff01000000 retries left: 3, oom retries left: 0, task failed due to oom: 1

    /tmp/ray/session_latest/logs/python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_60002.log:[2022-10-12 16:14:18,843 I 60002 60031] task_manager.cc:466: No retries left for task c8ef45ccd0112571ffffffffffffffffffffffff01000000, not going to resubmit.

.. note::

>     Task retries are executed immediately. If there is a long running process it is possible for a task to keep retrying and fail and exhaust the oom retries. Consider increasing ``RAY_task_oom_retries``, or :ref:`limit the number of concurrently running tasks <core-patterns-limit-running-tasks>`.

Reviewer: Is it something we will improve in the future?

Author: > Is it something we will improve in the future?

Yes but didn't want to make a hard commitment in the docs

.. note::

>     Actors by default are non-retriable since :ref:`max_restarts <actor-fault-tolerance>` defaults to 0, therefore tasks are preferred to actors when it comes to what gets killed first. Actor right now doesn't use ``RAY_task_oom_retries`` and instead uses :ref:`max_restarts <actor-fault-tolerance>` when it is killed by the memory monitor. This is to be changed very soon.

Reviewer 2: ```suggestion
    Actors by default are not retriable since :ref:`max_restarts <actor-fault-tolerance>` defaults to 0, therefore tasks are preferred to actors when it comes to what gets killed first. Actors currently don't use ``RAY_task_oom_retries`` and instead use :ref:`max_restarts <actor-fault-tolerance>` when killed by the memory monitor.
```

.. dropdown:: Example: memory monitor prefers to kill a retriable task

    Let's first start ray and specify the memory threshold.

    .. code-block:: bash

        RAY_memory_monitor_interval_ms=100 RAY_memory_usage_threshold_fraction=0.4 RAY_min_memory_free_bytes=-1 ray start --head

    Let's create an application two_actors.py that submits two actors, where the first one is retriable and the second one is non-retriable.

    .. literalinclude:: ../doc_code/ray_oom_prevention.py
          :language: python
          :start-after: __two_actors_start__
          :end-before: __two_actors_end__

    Run the application to see that only the first actor was killed.

    .. code-block:: bash

        $ python two_actors.py
        
        first actor was killed by memory monitor
        finished second actor

Addressing memory issues
------------------------

> When the application fails due to OOM, consider increasing the memory capacity of the node or :ref:`limit the number of concurrently running tasks <core-patterns-limit-running-tasks>`.

Reviewer: Can you also mention to inspect individual actor/task memory usage and reduce it?
